test/blackbox-tests/test-cases/variables-for-artifacts/run.t

```diff
@@ -0,0 +1,123 @@
+  $ dune clean
+  $ dune build --display short @t1
+      ocamldep .a1.objs/a.ml.d
+        ocamlc .a1.objs/byte/a1.{cmi,cmo,cmt}
+        ocamlc .a1.objs/byte/a1__A.{cmi,cmo,cmt}
+      ocamldep .b.eobjs/b.ml.d
+        ocamlc .b.eobjs/byte/b.{cmi,cmo,cmt}
+      ocamldep .c1.objs/c.ml.d
+        ocamlc .c1.objs/byte/c.{cmi,cmo,cmt}
+
+  $ dune build --verbose %{cmo:a} %{cmo:b} %{cmo:c} 2>&1 | grep -A100 'Actual targets'
+  # Actual targets:
+  # - _build/default/.a1.objs/byte/a1__A.cmo
+  # - _build/default/.b.eobjs/byte/b.cmo
+  # - _build/default/.c1.objs/byte/c.cmo
+
+  $ dune clean
+  $ dune build --display short @t2
+      ocamldep .a1.objs/a.ml.d
+        ocamlc .a1.objs/byte/a1.{cmi,cmo,cmt}
+        ocamlc .a1.objs/byte/a1__A.{cmi,cmo,cmt}
+
+  $ dune build --verbose %{cmi:a} 2>&1 | grep -A100 'Actual targets'
+  # Actual targets:
+  # - _build/default/.a1.objs/byte/a1__A.cmi
+
+  $ dune clean
+  $ dune build --display short @t3
+  File "dune", line 26, characters 0-38:
+  26 | (alias
+  27 |  (name t3)
+  28 |  (deps %{cmo:xxxx}))
+  Error: No rule found for xxxx.cmo
```

Comment:
When the object being referred to (here, a module named `xxxx` in the current directory), the macro expands anyway to a fictitious filename (here, `xxxx.cmo`) which then fails but with a message that is a bit misleading and does not mention the initial macro.

The problem is that it is not currently possible (as far as I can see) to propagate an error that occurs during expansion of a macro.

Reply:
Do you mean to propagate the location?

Reply:
nervermind, I see what you mean. Currently, the best is to error out while expanding the macro

Reply:
Actually, yh, we do need to allow reporting proper error when expanding macros. The result of `expand_var` is actually a bit misleading. We return `Error <expansion>` for things that need to be expanded later. Instead of returning a `result`, we should return something like this:

```
type expanded = Static of Value.t list | Dynamic of Pform.Expansion.t | Error of User_message.t
```

Reply:
I am trying to see how far the error case needs to be propagated inside the existing infrastructure. 

Does it sound reasonable that the result of `String_with_vars.expand` would be a `(Value.t list, User_message.t) result` or maybe even `(Value.t, User_message.t) result list` ?

Reply:
That seems right. We could also use the `Or_exn.t` type here since we use it for the same purpose in other places. Not too sure which of the two types you propose is better though.

Reply:
Thanks for the hint, will look into it.

Reply:
I think this is now fixed, thanks.